flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectResultFetcher.java

```diff
@@ -0,0 +1,359 @@
+ */
+public class CollectResultFetcher<T> {
+
+	private static final int DEFAULT_RETRY_MILLIS = 100;
+	private static final long DEFAULT_ACCUMULATOR_GET_MILLIS = 10000;
+
+	private static final Logger LOG = LoggerFactory.getLogger(CollectResultFetcher.class);
+
+	private final CompletableFuture<OperatorID> operatorIdFuture;
+	private final String accumulatorName;
+	private final int retryMillis;
+
+	private ResultBuffer buffer;
+
+	@Nullable
+	private JobClient jobClient;
+	@Nullable
+	private CoordinationRequestGateway gateway;
+
+	private boolean jobTerminated;
+	private boolean closed;
+
+	public CollectResultFetcher(
+			CompletableFuture<OperatorID> operatorIdFuture,
+			TypeSerializer<T> serializer,
+			String accumulatorName) {
+		this(
+			operatorIdFuture,
+			serializer,
+			accumulatorName,
+			DEFAULT_RETRY_MILLIS);
+	}
+
+	CollectResultFetcher(
+			CompletableFuture<OperatorID> operatorIdFuture,
+			TypeSerializer<T> serializer,
+			String accumulatorName,
+			int retryMillis) {
+		this.operatorIdFuture = operatorIdFuture;
+		this.accumulatorName = accumulatorName;
+		this.retryMillis = retryMillis;
+
+		this.buffer = new ResultBuffer(serializer);
+
+		this.jobTerminated = false;
+		this.closed = false;
+	}
+
+	public void setJobClient(JobClient jobClient) {
+		Preconditions.checkArgument(
+			jobClient instanceof CoordinationRequestGateway,
+			"Job client must be a CoordinationRequestGateway. This is a bug.");
+		this.jobClient = jobClient;
+		this.gateway = (CoordinationRequestGateway) jobClient;
+	}
+
+	@SuppressWarnings("unchecked")
+	public T next() {
+		if (closed) {
+			return null;
+		}
+
+		T res = buffer.next();
+		if (res != null) {
+			// we still have user-visible results, just use them
+			return res;
+		} else if (jobTerminated) {
+			// no user-visible results, but job has terminated, we have to return
+			return null;
+		}
+
+		// we're going to fetch some more
+		while (true) {
+			if (isJobTerminated()) {
+				// job terminated, read results from accumulator
+				jobTerminated = true;
+				Tuple2<Long, CollectCoordinationResponse> accResults = getAccumulatorResults();
+				if (accResults != null) {
+					buffer.dealWithResponse(accResults.f1, accResults.f0);
+				}
+				buffer.complete();
+			} else {
+				// job still running, try to fetch some results
+				CollectCoordinationResponse<T> response;
+				try {
+					response = sendRequest(buffer.version, buffer.offset);
+				} catch (Exception e) {
+					LOG.warn("An exception occurs when fetching query results", e);
+					sleepBeforeRetry();
+					continue;
+				}
+				buffer.dealWithResponse(response);
+			}
+
+			// try to return results after fetching
+			res = buffer.next();
+			if (res != null) {
+				// ok, we have results this time
+				return res;
+			} else if (jobTerminated) {
+				// still no results, but job has terminated, we have to return
+				return null;
+			} else {
```

Comment:
change this to `do {} while`, maybe we can reuse these statements with #110 - #117?

Reply:
This will cause iterator to sleep before the first try, which is undesired.